### PR TITLE
introduce new validation approach

### LINKED ISSUE #39: Change interface of run and validate to accept the data loader
See #37 for discussion

### LINKED ISSUE #42: add a validate handler that wraps trainer.validate to run on intervals and clear history
See #37 for details

### PR DESCRIPTION
**Pull request by jasonkriss**
Resolves #39 and resolves #42 

Need to sort out things in #43 before we merge this.

@alykhantejani how do you feel about the usage of a class here for the `Validate` handler? Imo, this is easier to follow than passing the arguments to `Trainer#add_event_handler` and having it forward them along. It's also nice for decomposing the functionality into multiple methods.

### CODE REVIEW

**Comment by jasonkriss** [comment]
@alykhantejani you mind if we get this one merged first? That way i can work against the new Trainer constructor while making changes in #43

**Comment by alykhantejani** [comment]
@jasonkriss sure, I have to run now but will take a look first thing tomorrow

**Comment by jasonkriss** [comment]
@alykhantejani 👍 thanks!

File: tests/ignite/handlers/test_validate.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from pytest import raises
+
+from ignite.handlers import Validate
+
+
+def test_validate_init():
+    Validate([], iteration_interval=1)
```
Comment (alykhantejani) [inline tests/ignite/handlers/test_validate.py]: I think these lines can be removed as they are not asserting anything

File: tests/ignite/trainer/test_trainer.py [hunk 2]
```diff
@@ -274,19 +264,17 @@ def end_of_iteration_handler(trainer):
 def test_terminate_after_training_iteration_skips_validation_run():
```
Comment (alykhantejani) [inline tests/ignite/trainer/test_trainer.py]: I guess this test can be removed as we now test above that `test_validate_is_not_called_by_default`

**Comment by alykhantejani** [comment]
Thanks @jasonkriss !

**Comment by jasonkriss** [comment]
Thanks @alykhantejani ! I'll try to wrap up #43 today.

**Comment by alykhantejani** [comment]
Thanks - no rush :)

**Comment by versatran01** [comment]
I'm a bit confused by the current validation approach. 
Where are we supposed to call the trainer.validate function? 
Are we supposed to put that into an event callback just call it there?
Also, the current example is lagging behind, I'd love to help to fix it but am confused for now.

**Comment by jasonkriss** [comment]
@versatran01 I apologize for the confusion. That is my fault for letting the example (and README) fall behind the current code. I plan on bringing everything up to date very soon. In the meantime, the  recommended approach is to use the [Validate](https://github.com/pytorch/ignite/blob/master/ignite/handlers/validate.py) handler and attach it to either `TRAINING_ITERATION_COMPLETED` or `TRAINING_EPOCH_COMPLETED`.

For example,

```python
from torch.utils.data import DataLoader
from torch import nn
import torch.nn.functional as F
from torch.optim import SGD
from torchvision.datasets import MNIST
from torchvision.transforms import Compose, ToTensor, Normalize
from ignite.trainer import TrainingEvents, create_supervised
from ignite.handlers import Validate
from ignite.metrics import categorical_accuracy

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x)
    
data_transform = Compose([ToTensor(), Normalize((0.1307,), (0.3081,))])

train_loader = DataLoader(MNIST(download=True, root=".", transform=data_transform, train=True),
                          batch_size=128, shuffle=True)

val_loader = DataLoader(MNIST(download=False, root=".", transform=data_transform, train=False),
                        batch_size=1024, shuffle=False)

model = Net().cuda()
optimizer = SGD(model.parameters(), lr=0.01, momentum=0.5)
trainer = create_supervised(model, optimizer, F.nll_loss, cuda=True)
trainer.add_event_handler(TrainingEvents.TRAINING_EPOCH_COMPLETED, Validate(val_loader, epoch_interval=1))

@trainer.on(TrainingEvents.TRAINING_EPOCH_COMPLETED)
def print_metrics(trainer):
    print(trainer.current_epoch, categorical_accuracy(trainer.validation_history))
    
trainer.run(train_loader, max_epochs=10)
```

**Comment by versatran01** [comment]
Thanks! This example is really helpful.

**Pull request merged**